File: fluss-flink/fluss-flink-common/src/main/java/com/alibaba/fluss/flink/tiering/source/split/TieringSplit.java
/*
 *  Copyright (c) 2025 Alibaba Group Holding Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.alibaba.fluss.flink.tiering.source.split;

import com.alibaba.fluss.metadata.TableBucket;
import com.alibaba.fluss.metadata.TablePath;

import org.apache.flink.api.connector.source.SourceSplit;

import javax.annotation.Nullable;

import java.util.Objects;

/** The table split for tiering service. It's used to describe the log data of a table bucket. */
public class TieringSplit implements SourceSplit {

    private static final String TIERING_SPLIT_PREFIX = "tiering-split-";

    private final TablePath tablePath;
    protected final TableBucket tableBucket;
    @Nullable protected final String partitionName;
    private final long startingOffset;
    private final long stoppingOffset;

    public TieringSplit(
            TablePath tablePath,
            TableBucket tableBucket,
            @Nullable String partitionName,
            long startingOffset,
            long stoppingOffset) {
        this.tablePath = tablePath;
        this.tableBucket = tableBucket;
        this.partitionName = partitionName;
        this.startingOffset = startingOffset;
        this.stoppingOffset = stoppingOffset;
        if ((tableBucket.getPartitionId() == null && partitionName != null)
                || (tableBucket.getPartitionId() != null && partitionName == null)) {
            throw new IllegalArgumentException(
                    "Partition name and partition id must be both null or both not null.");
        }
    }

    public long getStartingOffset() {
        return startingOffset;
    }

    public long getStoppingOffset() {
        return stoppingOffset;
    }

    public TablePath getTablePath() {
        return tablePath;
    }

    public TableBucket getTableBucket() {
        return tableBucket;
    }

    @Nullable
    public String getPartitionName() {
        return partitionName;
    }

    public String splitId() {
        if (tableBucket.getPartitionId() != null) {
            return TIERING_SPLIT_PREFIX
                    + tableBucket.getTableId()
                    + "-p"
                    + tableBucket.getPartitionId()
                    + "-"
                    + tableBucket.getBucket();
        } else {
            return TIERING_SPLIT_PREFIX + tableBucket.getTableId() + "-" + tableBucket.getBucket();
        }
    }

    @Override
    public String toString() {
        return "TieringSplit{"
                + "tablePath="
                + tablePath
                + ", tableBucket="
                + tableBucket
                + ", partitionName='"
                + partitionName
                + '\''
                + ", startingOffset="
                + startingOffset
                + ", stoppingOffset="
                + stoppingOffset
                + '}';
    }

    @Override
    public boolean equals(Object object) {
        if (!(object instanceof TieringSplit)) {
            return false;
        }
        TieringSplit that = (TieringSplit) object;
        return startingOffset == that.startingOffset
                && stoppingOffset == that.stoppingOffset
                && Objects.equals(tablePath, that.tablePath)
                && Objects.equals(tableBucket, that.tableBucket)
                && Objects.equals(partitionName, that.partitionName);
    }

    @Override
    public int hashCode() {
        return Objects.hash(tablePath, tableBucket, partitionName, startingOffset, stoppingOffset);
    }
}

File: fluss-spark/fluss-spark-common/src/main/scala/org/apache/fluss/spark/tiering/TieringSplit.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.fluss.spark.tiering

import org.apache.fluss.metadata.{TableBucket, TablePath}

/** Sealed trait representing a tiering split for distributing work across Spark executors. */
sealed trait TieringSplit extends Serializable {
  def tablePath: TablePath
  def tableBucket: TableBucket
  def partitionName: Option[String]
  def numberOfSplits: Int
}

/**
 * A split for reading log data from Fluss.
 *
 * @param tablePath
 *   the table path
 * @param tableBucket
 *   the table bucket
 * @param partitionName
 *   optional partition name
 * @param startingOffset
 *   the starting log offset (inclusive)
 * @param stoppingOffset
 *   the stopping log offset (exclusive)
 * @param numberOfSplits
 *   total number of splits in this tiering round
 */
case class TieringLogSplit(
    tablePath: TablePath,
    tableBucket: TableBucket,
    partitionName: Option[String],
    startingOffset: Long,
    stoppingOffset: Long,
    numberOfSplits: Int)
  extends TieringSplit

/**
 * A split for reading a KV snapshot from Fluss (used for primary key tables on first tiering).
 *
 * @param tablePath
 *   the table path
 * @param tableBucket
 *   the table bucket
 * @param partitionName
 *   optional partition name
 * @param snapshotId
 *   the KV snapshot ID to read
 * @param logOffsetOfSnapshot
 *   the log offset corresponding to the snapshot
 * @param numberOfSplits
 *   total number of splits in this tiering round
 */
case class TieringSnapshotSplit(
    tablePath: TablePath,
    tableBucket: TableBucket,
    partitionName: Option[String],
    snapshotId: Long,
    logOffsetOfSnapshot: Long,
    numberOfSplits: Int)
  extends TieringSplit
